diff -Nur linux-2.6.32.33/drivers/mmc/host/au1xmmc.c linux-2.6.32.33-loongson2f/drivers/mmc/host/au1xmmc.c
--- linux-2.6.32.33/drivers/mmc/host/au1xmmc.c	2011-03-15 05:30:16.000000000 +0800
+++ linux-2.6.32.33-loongson2f/drivers/mmc/host/au1xmmc.c	2011-03-26 03:39:49.000000000 +0800
@@ -650,11 +650,11 @@
 				flags = DDMA_FLAGS_IE;
 
 			if (host->flags & HOST_F_XMIT) {
-				ret = au1xxx_dbdma_put_source_flags(channel,
-					(void *)sg_virt(sg), len, flags);
+				ret = au1xxx_dbdma_put_source(channel,
+					sg_phys(sg), len, flags);
 			} else {
-				ret = au1xxx_dbdma_put_dest_flags(channel,
-					(void *)sg_virt(sg), len, flags);
+				ret = au1xxx_dbdma_put_dest(channel,
+					sg_phys(sg), len, flags);
 			}
 
 			if (!ret)
