diff -Nru linux-2.6.36/sound/soc/s3c24xx/Kconfig linux-2.6.36-Firendyarm/sound/soc/s3c24xx/Kconfig
--- linux-2.6.36/sound/soc/s3c24xx/Kconfig	2010-10-20 21:30:22.000000000 +0100
+++ linux-2.6.36-Firendyarm/sound/soc/s3c24xx/Kconfig	2010-10-26 08:47:51.000000000 +0100
@@ -126,6 +126,15 @@
 	help
 	  Sat Y if you want to add support for SoC audio on the SMDK.
 
+config SND_SOC_MINI_WM9713
+	tristate "SoC AC97 Audio support for Mini6410 with WM9713"
+	depends on SND_S3C24XX_SOC && MACH_MINI6410
+	select SND_SOC_WM9713
+	select SND_S3C_SOC_AC97
+	help
+	  Say Y if you want to add support for SoC audio on the FriendlyARM
+	  Mini6410 board.
+
 config SND_S3C64XX_SOC_SMARTQ
 	tristate "SoC I2S Audio support for SmartQ board"
 	depends on SND_S3C24XX_SOC && MACH_SMARTQ
diff -Nru linux-2.6.36/sound/soc/s3c24xx/Makefile linux-2.6.36-Firendyarm/sound/soc/s3c24xx/Makefile
--- linux-2.6.36/sound/soc/s3c24xx/Makefile	2010-10-20 21:30:22.000000000 +0100
+++ linux-2.6.36-Firendyarm/sound/soc/s3c24xx/Makefile	2010-10-26 08:47:51.000000000 +0100
@@ -29,6 +29,7 @@
 snd-soc-s3c24xx-simtec-tlv320aic23-objs := s3c24xx_simtec_tlv320aic23.o
 snd-soc-smdk64xx-wm8580-objs := smdk64xx_wm8580.o
 snd-soc-smdk-wm9713-objs := smdk_wm9713.o
+snd-soc-mini6410-wm9713-objs := mini6410_wm9713.o
 snd-soc-s3c64xx-smartq-wm8987-objs := smartq_wm8987.o
 
 obj-$(CONFIG_SND_S3C24XX_SOC_JIVE_WM8750) += snd-soc-jive-wm8750.o
@@ -42,4 +43,5 @@
 obj-$(CONFIG_SND_S3C24XX_SOC_SIMTEC_TLV320AIC23) += snd-soc-s3c24xx-simtec-tlv320aic23.o
 obj-$(CONFIG_SND_S3C64XX_SOC_WM8580) += snd-soc-smdk64xx-wm8580.o
 obj-$(CONFIG_SND_SOC_SMDK_WM9713) += snd-soc-smdk-wm9713.o
+obj-$(CONFIG_SND_SOC_MINI_WM9713) += snd-soc-mini6410-wm9713.o
 obj-$(CONFIG_SND_S3C64XX_SOC_SMARTQ) += snd-soc-s3c64xx-smartq-wm8987.o
 
