--- ./arch/arm/plat-samsung/include/plat/map-base.h	2010-09-29 09:09:08.000000000 +0800
+++ /mnt/hgfs/share/openWRT/linux-2.6.35.7-linux-smdkv210/arch/arm/plat-samsung/include/plat/map-base.h	2011-08-08 12:52:05.000000000 +0800
@@ -22,7 +22,7 @@
  * an single MOVS instruction (ie, only 8 bits of set data)
  */
 
-#define S3C_ADDR_BASE	(0xF4000000)
+#define S3C_ADDR_BASE	(0xF4000000)
 
 #ifndef __ASSEMBLY__
 #define S3C_ADDR(x)	((void __iomem __force *)S3C_ADDR_BASE + (x))
@@ -35,6 +35,8 @@
 #define S3C_VA_MEM	S3C_ADDR(0x00200000)	/* memory control */
 #define S3C_VA_TIMER	S3C_ADDR(0x00300000)	/* timer block */
 #define S3C_VA_WATCHDOG	S3C_ADDR(0x00400000)	/* watchdog */
+#define S3C_VA_OTG	S3C_ADDR(0x00E00000)	/* OTG */
+#define S3C_VA_OTGSFR	S3C_ADDR(0x00F00000)	/* OTG PHY */
 #define S3C_VA_UART	S3C_ADDR(0x01000000)	/* UART */
 
 /* This is used for the CPU specific mappings that may be needed, so that
