#
# modify mach-smdk2440.c
#
--- linux-2.6.32.27/arch/arm/mach-s3c2440/mach-smdk2440.c.orig	2011-04-03 05:26:03.000000000 -0400
+++ linux-2.6.32.27/arch/arm/mach-s3c2440/mach-smdk2440.c	2011-04-03 05:26:03.000000000 -0400
@@ -173,7 +173,7 @@
 	smdk_machine_init();
 }
 
-MACHINE_START(S3C2440, "SMDK2440")
+MACHINE_START(SMDK2440, "SMDK2440")
 	/* Maintainer: Ben Dooks <ben@fluff.org> */
 	.phys_io	= S3C2410_PA_UART,
 	.io_pg_offst	= (((u32)S3C24XX_VA_UART) >> 18) & 0xfffc,
#
# add fs2410, gec2410 in mach-s3c2410/
#
--- linux-2.6.32.27/arch/arm/mach-s3c2410/Kconfig.orig	2012-03-17 18:14:52.000000000 +0800
+++ linux-2.6.32.27/arch/arm/mach-s3c2410/Kconfig	2012-08-28 10:59:30.938056228 +0800
@@ -162,4 +162,20 @@
 	help
 	   Say Y here if you are using the Armzone QT2410
 
+config MACH_FS2410
+	bool "FS2410"
+	select CPU_S3C2410
+	select S3C_DEV_USB_HOST
+	select S3C_DEV_NAND
+	help
+	   Say Y here if you are using the Ucdragon FS2410
+
+config MACH_GEC2410
+	bool "GEC2410"
+	select CPU_S3C2410
+	select S3C_DEV_USB_HOST
+	select S3C_DEV_NAND
+	help
+	   Say Y here if you are using the GEC2410
+
 endmenu
--- linux-2.6.32.27/arch/arm/mach-s3c2410/Makefile.orig	2012-03-17 18:14:52.000000000 +0800
+++ linux-2.6.32.27/arch/arm/mach-s3c2410/Makefile	2012-08-28 10:59:30.938056228 +0800
@@ -31,6 +31,8 @@
 obj-$(CONFIG_MACH_TCT_HAMMER)	+= mach-tct_hammer.o
 obj-$(CONFIG_MACH_VR1000)	+= mach-vr1000.o usb-simtec.o
 obj-$(CONFIG_MACH_QT2410)	+= mach-qt2410.o
+obj-$(CONFIG_MACH_GEC2410)	+= mach-gec2410.o
+obj-$(CONFIG_MACH_FS2410)	+= mach-fs2410.o
 
 # Common bits of machine support
 
#
# add qq2440, gec2440 in mach-s3c2440/
#
--- linux-2.6.32.27/arch/arm/mach-s3c2440/Kconfig.orig	2012-03-17 18:14:52.000000000 +0800
+++ linux-2.6.32.27/arch/arm/mach-s3c2440/Kconfig	2012-08-28 10:59:31.042056226 +0800
@@ -99,14 +99,33 @@
 config MACH_MINI2440
 	bool "MINI2440 development board"
 	select CPU_S3C2440
-	select EEPROM_AT24
-	select LEDS_TRIGGER_BACKLIGHT
-	select SND_S3C24XX_SOC_S3C24XX_UDA134X
 	select S3C_DEV_NAND
 	select S3C_DEV_USB_HOST
+	select S3C2440_XTAL_12000000
 	help
 	  Say Y here to select support for the MINI2440. Is a 10cm x 10cm board
 	  available via various sources. It can come with a 3.5" or 7" touch LCD.
 
+config MACH_QQ2440
+	bool "QQ2440 development board"
+	select CPU_S3C2440
+	select S3C_DEV_NAND
+	select S3C_DEV_USB_HOST
+	select S3C2440_XTAL_12000000
+	help
+	  Say Y here to select support for the QQ2440. Is a 12cm x 10cm board
+	  available via various sources. It can come with a 3.5" or 7" touch LCD.
+
+config MACH_GEC2440
+	bool "GEC2440 development board"
+	select CPU_S3C2440
+	select S3C_DEV_NAND
+	select S3C_DEV_USB_HOST
+	select S3C2440_XTAL_12000000
+	help
+	  Say Y here to select support for the GEC2440. Is a 17cm x 17cm board
+	  available via various sources. It can come with a 3.5" touch LCD.
+
+
 endmenu
 
--- linux-2.6.32.27/arch/arm/mach-s3c2440/Makefile.orig	2012-03-17 18:14:52.000000000 +0800
+++ linux-2.6.32.27/arch/arm/mach-s3c2440/Makefile	2012-08-28 10:59:31.042056226 +0800
@@ -23,3 +23,5 @@
 obj-$(CONFIG_MACH_NEXCODER_2440) += mach-nexcoder.o
 obj-$(CONFIG_MACH_AT2440EVB) += mach-at2440evb.o
 obj-$(CONFIG_MACH_MINI2440) += mach-mini2440.o
+obj-$(CONFIG_MACH_QQ2440) += mach-qq2440.o
+obj-$(CONFIG_MACH_GEC2440) += mach-gec2440.o
