Index: linux-2.6.23-rc8/drivers/serial/s3c2410.c
===================================================================
--- linux-2.6.23-rc8.orig/drivers/serial/s3c2410.c
+++ linux-2.6.23-rc8/drivers/serial/s3c2410.c
@@ -708,10 +708,6 @@
 		int calc_deviation;
 
 		for (sptr = res; sptr < resptr; sptr++) {
-			printk(KERN_DEBUG
-			       "found clk %p (%s) quot %d, calc %d\n",
-			       sptr->clksrc, sptr->clksrc->name,
-			       sptr->quot, sptr->calc);
 
 			calc_deviation = baud - sptr->calc;
 			if (calc_deviation < 0)
@@ -723,12 +719,8 @@
 			}
 		}
 
-		printk(KERN_DEBUG "best %p (deviation %d)\n", best, deviation);
 	}
 
-	printk(KERN_DEBUG "selected clock %p (%s) quot %d, calc %d\n",
-	       best->clksrc, best->clksrc->name, best->quot, best->calc);
-
 	/* store results to pass back */
 
 	*clksrc = best->clksrc;
